feat(tactic/suggest): generalize and reimplement library_search

u6338192 opened the pull request:
TO CONTRIBUTORS:

Make sure you have:

  * [x] for tactics:
     * [x] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [x] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)


If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

This PR is for the `suggest` tactic. `suggest` gives a list of applicable lemmas. The output is of the form `exact [lemma name] ...` or `refine [lemma name]...`. (It accepts an optional argument limiting the number of lemmas to list.)

As well as adding src/suggest.lean and test/suggest.lean files, this PR also adds some documentation about `suggest` to doc/tactics.md and modifies the code for `library_search` (but not the behaviour) in the following ways:

* Added `replace_mvars` and `tactic_statement` functions and modified the main `library_search` function to use these.
* Modified the `apply_and_solve` function to take a new parameter `close_goals`. If `close_goals` is false, `apply_and_solve` will try to fill in as many holes as possible (even if this isn't all of them) using `solve_by_elim`.
* Modified the `apply_declaration` function to take a new parameter `close_goals`, which is passed to `apply_and_solve`. Also `apply_and_solve` has been refactored to make it more readable.

Note that the modifications to `apply_and_solve` and `apply_declaration` were done by Scott Morrison. These changes are so that `library_search` and `suggest` can share code. They don't change the behaviour of `library_search`.

Commit:
Merge pull request #1 from leanprover-community/master

Updating fork u6338192

Commit:
Merge pull request #2 from leanprover-community/master

u6338192 update fork

Commit:
Merge pull request #3 from leanprover-community/master

Updating my fork baby

Commit:
Create refine_list.lean

The refine_list tactic.

Commit:
Create refine_list.lean

The refine_list.lean file

Commit:
Update tactics.md

Added refine_list docs.

Commit:
Update library_search.lean

Added replace_mvars function

Commit:
Update library_search.lean

Added tactic_statement function and some code in the main library_search function which uses it.

Commit:
commits

Commit:
refine_list commits

Commit:
Scott's work

Commit:
update tactics.md

Commit:
doc strings

Commit:
Merge branch 'refine_list' of github.com:u6338192/mathlib into refine_list

Commit:
update

Commit:
Merge branch 'refine_list' of https://github.com/u6338192/mathlib into refine_list

jcommelin commented on src/tactic/refine_list.lean:
```diff
@@ -0,0 +1,118 @@
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Lucas Allen
+
+-- TODO Documentation in tactics.md
```
This is done, right?

jcommelin commented on src/tactic/refine_list.lean:
```diff
@@ -0,0 +1,118 @@
+declare_trace silence_refine_list -- Turn off `exact ...` trace message
+declare_trace refine_list         -- Trace a list of all relevant lemmas
+
+/-- The main list_refine tactic, this is very similar to the main library_search function -/
```
Remember that docstrings are most helpful when they show a message in a tooltip that is meaningful outside the context of this file.
Of course this is not the interactive tactic. I see that you have a very nice docstring over there. It's not so important, but I think this docstring could be improved by giving a very short (1 or 2 lines) summary of what this tactic does.
Finally, the docstring says `list_refine` instead of `refine_list`.

jcommelin replied:
Yeah, looks good to me.

fpvandoorn commented on src/tactic/refine_list.lean:
```diff
@@ -0,0 +1,118 @@
+open tactic.library_search
+
+/-- Returns the current tactic_state. -/
+meta def get_state : tactic tactic_state :=
```
This is called `tactic.read`

fpvandoorn commented on src/tactic/refine_list.lean:
```diff
@@ -0,0 +1,118 @@
+λ s, result.success s s
+
+/-- Sets the tactic_state. -/
+meta def set_state (s' : tactic_state) : tactic tactic_state :=
```
This can be replaced by `tactic.write` in your applications below, right?

fpvandoorn commented on src/tactic/refine_list.lean:
```diff
@@ -0,0 +1,118 @@
+   -- Turn defs into an mllist
+   let mldefs := get_mldefs defs,
+   -- Filter out the lemmas that cannot be used with refine
+   let results := mldefs.mfilter_map
```
Would it be more efficient to compute `results_with_num_goals` immediately? That would also eliminate the need for `run_and_save_state`.

Commit:
Update refine_list.lean

Removes stray TODO line and changed the docstring for the main refine_list function.

Commit:
Update refine_list.lean

Spelling...

Commit:
get_state and set_state replace by read and write

Commit:
Combined functions and removed run_and_save_state

Commit:
removed symmetry from library_search

Commit:
removing test that doesn't work without symmetry

Commit:
test properly

Commit:
rename `refine_list` to `suggest`

Commit:
cleaning up

Commit:
minor

Commit:
better sorting of results

Commit:
oops, turn off trace messages again

Commit:
type annotation

Commit:
optimisation, and simpler logic

Commit:
further explanation

Commit:
a little more cleanup

robertylewis commented on src/tactic/library_search.lean:
```diff
@@ -76,29 +76,62 @@ meta def library_defs (hs : name) : tactic (list decl_data) :=
+/-- Fail if the target contains a metavariable. -/
+meta def no_mvars_in_target : tactic unit :=
+do t ← target,
+   guard t.list_meta_vars.empty
```
```suggestion
   guard $ bnot t.has_meta_var
```
The fun monad version is something like `expr.has_meta_var <$> target >>= guardb ∘ bnot`

kim-em replied:
Thanks. I've used this, and further improved the comments explaining why we use `no_mvars_in_target`.

robertylewis commented on src/tactic/library_search.lean:
```diff
@@ -76,29 +76,62 @@ meta def library_defs (hs : name) : tactic (list decl_data) :=
+-- we first run `solve_by_elim` separately on a subset of the goals, whether or not `close_goals` is set,
+-- and then if `close_goals = tt`, require that `solve_by_elim { all_goals := tt }` succeeds
+-- on the remaining goals.
+meta def apply_and_solve (close_goals : bool) (discharger : tactic unit) (e : expr) : tactic unit :=
```
I think the goal handling is off here. `

If there's more than one goal to start with, the `any_goals` call will touch them. `apply_and_solve` is only supposed to act on the first goal on the (original) stack, right? The whole thing should be wrapped in `focus1`.

Also -- maybe I'm being dense here, it's late -- are lines 95-100 the same as 
```lean
all_goals {try {solve_by_elim {discharger := discharger}}} >> (done <|> guard ¬ close_goals)`
```
? I'm not seeing why the `propositional_goal` and `no_mvars_in_target` are doing anything.

kim-em replied:
I'll add some further explanation, but let me try here first.

`solve_by_elim` is happy to solve a non-singleton goal using a local hypothesis (e.g. fill in a `nat`), even if some other goal depends on it. Thus calling `solve_by_elim` separately on each goal can cause it to fill in a bad solution to one goal, without the possibility of backtracking later.
The mode `solve_by_elim { all_goals := tt }` performs backtracking across multiple goals. However, `solve_by_elim` fails unless it discharges all the goals it's been given.

What behaviour do we want? For `suggest`, we would like "as many subgoals as possible" discharged for us by `solve_by_elim` after the application of a library lemma. The current code attempts to first discharge any "safe" goals, i.e. purely propositional goals not containing metavariables. (Other goals can't depend on the solution we pick for a propositional goal, by proof irrelevance, unless perhaps the goal involves a metavariable. As an example, if you try to prove `a = b` by apply `eq.trans`, the first goal is `a = ?m`, which looks propositional so `solve_by_elim` would happily solve using `rfl`, which is likely a mistake.)

After we've attempted discharging the "safe" goals, we then do a second pass and try to solve all the goals together (with backtracking). If `close_goals = tt`, we insist that this second pass succeeds. This step allows us to fill in "data" solutions for goals, as long as we subsequently solve all the goals.

There is a potential performance problem here: the "safe" goals will (likely) be considered _twice_ by `solve_by_elim`. (In fact, if there are unsafe goals, in the second pass `solve_by_elim` will work on those first, just because they are typically first in the list of goals, and not actually get to the safe goals unless it has an idea for the unsafe goals.) However if we only have "safe" goals, but we can't solve them, the current implementation does try them twice, which is pretty unfortunate.

I *think* a satisfactory solution to this may be to check "does there exist a non-safe goal", and only run the second phase of `solve_by_elim` if this is the case. This should remove nearly all the duplication of effort.

kim-em replied:
Adding this test seems to work well. I've committed an update, that also includes a much more detailed discussion of the logic in `apply_and_solve`.

robertylewis replied:
Okay, thanks for clarifying, this makes much more sense now. I had assumed `solve_by_elim {all_goals := tt}` was the same as `all_goals {solve_by_elim}`, but that's not the case. This naming is slightly confusing.

robertylewis replied:
Regarding my first comment: I noticed `library_search` fails immediately unless there's exactly one goal, so `apply_and_solve` should never get multiple goals to start with, right? Then my comment doesn't matter for that case. But I think it doesn't hurt to put `focus1 $` at the very beginning of `apply_and_solve` just in case it ends up being used somewhere else. (If not, maybe note in the doc string that it assumes there's only one goal.)

kim-em replied:
I think I'll just fix `library_search` and `suggest` so that they don't mind being given multiple goals (by using `focus1`).

kim-em replied:
I wonder if there is a clearer name for the `all_goals` option in `solve_by_elim`? `backtrack_all_goals`, perhaps? Long-winded, but much clearer intent.

robertylewis commented on src/tactic/library_search.lean:
```diff
@@ -76,29 +76,62 @@ meta def library_defs (hs : name) : tactic (list decl_data) :=
+   g ← head_beta g,
+   r ← pp (replace_mvars g),
+   r ← pure $ if g.has_meta_var then sformat!"refine {r}" else sformat!"exact {r}",
+   return r
```
132 and 133 are just
```suggestion
   if g.has_meta_var then sformat!"refine {r}" else sformat!"exact {r}"
```

robertylewis commented on src/tactic/suggest.lean:
```diff
@@ -0,0 +1,104 @@
+   -- Sort by number of remaining goals, then by number of hypotheses used.
+   let L := L.qsort(λ d₁ d₂, d₁.2.1 < d₂.2.1 ∨ (d₁.2.1 = d₂.2.1 ∧ d₁.2.2 ≥ d₂.2.2)),
+   -- Print the first num successful lemmas
+   if L.length = 0 then do
```
```suggestion
   if L.length = 0 then 
```

robertylewis commented on src/tactic/suggest.lean:
```diff
@@ -0,0 +1,104 @@
+
+end tactic
+
+namespace interactive
```
This block should move up a few lines, before you close `tactic`. The namespace you need is `tactic.interactive`.

kim-em replied:
That's fine.

robertylewis commented on src/tactic/suggest.lean:
```diff
@@ -0,0 +1,104 @@
+`suggest` might miss some results if `num` is not large enough. However, because
+`suggest` uses monadic lazy lists, smaller values of `num` run faster than larger values.
+-/
+meta def suggest := tactic.suggest
```
```suggestion
meta def suggest : tactic unit := tactic.suggest >> tactic.skip
```
Interactive tactics should have type `tactic unit`.

robertylewis commented on test/suggest.lean:
```diff
@@ -0,0 +1,68 @@
+-- Copyright (c) 2019 Lucas Allen. All rights reserved.
```
Use `/- -/` for copyright headers.

robertylewis commented on src/tactic/suggest.lean:
```diff
@@ -0,0 +1,104 @@
+-- Copyright (c) 2019 Lucas Allen. All rights reserved.
```
Use `/- -/` for copyright headers.

robertylewis reviewed (commented):
I suspect that some of the behavior of `suggest` could be useful in other contexts too. Specifically, producing a list of lemmas that progress the goal is a generally interesting thing to do. Do you think you could separate this logic from the printing? I'd personally have `tactic.suggest` silently produce a list of lemma names (or patterns, or whatever's convenient), and then do the formatting directly in the interactive version.

Commit:
Update src/tactic/library_search.lean

Rob Lewis's refactoring of `tactic_statement`

Co-Authored-By: Rob Lewis <Rob.y.lewis@gmail.com>

Commit:
Update src/tactic/suggest.lean

Remove `do` in line 82

Co-Authored-By: Rob Lewis <Rob.y.lewis@gmail.com>

Commit:
Update src/tactic/suggest.lean

Rob's update of suggest in tactic.interactive.

Co-Authored-By: Rob Lewis <Rob.y.lewis@gmail.com>

Commit:
changed interactive to tactic.interactive

Commit:
Merge branch 'refine_list' of https://github.com/u6338192/mathlib into refine_list

Commit:
used /--/ for copyright headers

Commit:
Merge branch 'refine_list' of https://github.com/u6338192/mathlib into refine_list

Commit:
adding comments explaining logic in apply_and_solve, and an optimisation

Commit:
Merge branch 'refine_list' of github.com:u6338192/mathlib into refine_list

Commit:
refactor no_mvars_in_target

Commit:
add missing argument to interactive tactic

Commit:
move all printing to the interactive tactic

Commit:
refactoring

kim-em committed:
Merge branch 'master' into refine_list

Commit:
cleanup

robertylewis commented on src/tactic/library_search.lean:
```diff
@@ -76,29 +76,75 @@ meta def library_defs (hs : name) : tactic (list decl_data) :=
+Fail unless the goal is "safe", i.e. propositional, and does not include any metavariables.
+If a goal is "safe", discharging it does not affect the provability of any other goals.
+-/
+meta def safe : tactic unit :=
```
This and `no_mvars_in_target` can probably move to `tactic.core`, although `safe` should be renamed.

kim-em replied:
`tactic.core` already has the much more ambitious `terminal_goal`, which tries to establish that no other goals can depend on the value of the given goal (propositional and subsingleton goals are automatically terminal, and after that we search the types of other metavariables in the result for the current metavariable). I could just use that.

kim-em replied:
Done.

robertylewis commented on src/tactic/suggest.lean:
```diff
@@ -0,0 +1,116 @@
+* `message`, a string of the form `refine ...` or `exact ...` which will reproduce that tactic state.
+-/
+meta def suggest (limit : option ℕ := none) (discharger : tactic unit := done) : tactic (list (tactic_state × string)) :=
+do [g] ← get_goals | fail "`suggest` should be called with exactly one goal",
```
Is this restriction really necessary? Rather than require one goal, you could just focus on the first goal. (Same goes for `library_search`.) Most of the time we hope users will have done that already. But these tactics are very useful for beginners who might not be writing good structured proofs yet.

kim-em replied:
I fixed this in the big refactor that built `library_search` on top of `suggest_core`. Both tactics are now happy with multiple goals.

kim-em committed:
refactoring

kim-em committed:
complete refactor; library_search is defined in terms of suggest_core now

kim-em committed:
Merge remote-tracking branch 'origin/master' into refine_list

kim-em committed:
restore protect to list.traverse and option.traverse

robertylewis commented on src/tactic/suggest.lean:
```diff
@@ -0,0 +1,311 @@
+   return $ L.map application.script
+
+/--
+Returns a string of the form `exact ...`, which closes the current goal (and closes the goal).
```
```suggestion
Returns a string of the form `exact ...`, which closes the current goal.
```

robertylewis reviewed (commented):
Could you add doc strings for the defs at the top of `suggest.lean`? Otherwise I think this is good to go!

kim-em committed:
Update src/tactic/suggest.lean

Co-Authored-By: Rob Lewis <Rob.y.lewis@gmail.com>

kim-em committed:
some doc comments

kim-em committed:
Merge branch 'master' into refine_list
